metaserver/src/main/scala/scala/meta/languageserver/ScalametaLanguageServer.scala

```diff
@@ -189,58 +189,12 @@ class ScalametaLanguageServer(
 
   override def documentSymbols(
       td: TextDocumentIdentifier
-  ): Seq[SymbolInformation] = {
-    import scala.meta._
-
-    // For a given node returns its closest ancestor which is a definition, declaration or a package object
-    // NOTE: package is not considered a wrapping definition, but it could be (a subject to discuss)
-    def wrappingDefinition(t: Tree): Option[Tree] = {
-      if (t.is[Defn] ||
-        t.is[Decl] ||
-        t.is[Pkg.Object]) Some(t)
-      else t.parent.flatMap(wrappingDefinition)
-    }
-
-    def parentMember(t: Tree): Option[Tree] = {
-      if (t.is[Member.Term] || t.is[Member.Type]) Some(t)
-      else t.parent.flatMap(parentMember)
-    }
-
-    // This is needed only to unfold full package names
-    def qualifiedName(t: Tree): Option[String] = t match {
-      case Term.Name(name) =>
-        Some(name)
-      case Term.Select(qual, name) =>
-        qualifiedName(qual).map { prefix =>
-          s"${prefix}.${name}"
-        }
-      case Pkg(sel: Term.Select, _) =>
-        qualifiedName(sel)
-      case m: Member =>
-        Some(m.name.value)
-      case _ => None
-    }
-
+  ): List[SymbolInformation] = {
     val path = Uri.toPath(td.uri).get
-    val contents = buffers.read(path)
-    for {
-      tree <- contents.parse[Source].toOption.toList
-      node <- tree.collect {
-        case n if n.is[Member.Type] || n.is[Member.Term] => n
-      }
-      name <- qualifiedName(node)
-      // Package as a wrapping definition for itself:
-      defn <- if (node.is[Pkg]) Some(node) else wrappingDefinition(node)
-    } yield
-      SymbolInformation(
-        name,
-        defn.symbolKind,
-        path.toLocation(defn.pos),
-        defn.parent
-          .flatMap(parentMember)
-          .flatMap(wrappingDefinition)
-          .flatMap(qualifiedName)
-      )
+    buffers.source(path) match {
+      case Some(source) => OutlineProvider.documentSymbols(path, source)
```

Comment:
Thanks for moving this to the new *Provider structure, I was gonna nag you about that but you beat me to it and fixed it already 😄